### PR TITLE
Support wallet_watchAsset

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/17878

Image parameter is not supported yet because we currently don't support getting a logo from a remote URL, currently it will still use the placeholder icon if a remote URL is passed. (https://github.com/brave/brave-browser/issues/20000)

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Visit https://www.coingecko.com/en/coins/railgun and click on the fox icon to add suggest token
add_to_metamask

2. Prompt should show up for adding suggest token (will be asking to unlock first if wallet is locked)
add_suggest_token

3. Cancel the request.
4. Open a new tab and visit brave://wallet/crypto/portfolio should not see the new token.
5. Do step 1 again and accept the request this time.
6. Go to brave://wallet/crypto/portfolio should see it being listed in visible assets. (Existing page will need to fresh to see the change)
assets

7. Visit https://www.coingecko.com/en/coins/tether and click on the fox icon to add suggest token
8. Prompt should show up with correct logo.
usdt

 9. Accept the request.
 10. Open a new tab and visit brave://wallet/crypto/portfolio should see tether in the asset list.
Screen Shot 2021-12-08 at 5 33 17 PM

### CODE REVIEW

File: components/brave_wallet/browser/brave_wallet_service.cc [hunk 1]
```diff
@@ -692,4 +677,90 @@ void BraveWalletService::AddSignMessageRequest(
...
+      std::move(callback).Run(approved, mojom::ProviderError::kSuccess, "");
+    }
+  }
+}
+
+void BraveWalletService::OnNetworkChanged() {
+  add_suggest_token_requests_.clear();
+  // Reject pending suggest token requests when network changed.
+  for (auto& callback : add_suggest_token_callbacks_)
+    std::move(callback.second).Run(false, mojom::ProviderError::kSuccess, "");
```
Comment (bbondy) [inline components/brave_wallet/browser/brave_wallet_service.cc:762]: should it be `mojom::ProviderError::kUserRejectedRequest` here?

File: components/brave_wallet/common/eth_request_helper.cc [hunk 2]
```diff
@@ -365,4 +380,106 @@ bool ParseSwitchEthereumChainParams(const std::string& json,
...
+  if (decimals < 0 || decimals > 36) {
+    *error_message = base::StringPrintf(
+        "Invalid decimals '%d': decimals should be greater than 0 and less "
+        "than 36",
+        decimals);
+    return false;
+  }
+
+  // TODO(jocelyn): Parse image URL to be the logo URL once we supports
+  // downloading an icon from a remote URL.
```
Comment (bbondy) [inline components/brave_wallet/common/eth_request_helper.cc:476]: Let's parse it pls just so it's stored at least for the future when we want to use it. We can put it in `logo` which already exists.

File: components/brave_wallet/browser/brave_wallet_service.cc (hunk 1 above)
Comment (yrliou) [reply components/brave_wallet/browser/brave_wallet_service.cc]: https://github.com/brave/brave-core/pull/11518/commits/b795f4e65add1f1643a8f1e7aa91dd8120cd4dd6

File: components/brave_wallet/common/eth_request_helper.cc (hunk 2 above)
Comment (yrliou) [reply components/brave_wallet/common/eth_request_helper.cc]: https://github.com/brave/brave-core/pull/11518/commits/e8d4fe2f5e0ad9aa2989212b10d59932eea368b8

**Review by bbondy — APPROVED** [review]

File: components/brave_wallet_ui/utils/string-utils.ts [hunk 3]
```diff
@@ -4,3 +4,6 @@ export const stripERC20TokenImageURL = (url?: string) =>
 export const toProperCase = (value: string) =>
   value.replace(/\w\S*/g,
     (txt) => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase())
+
+export const isRemoteImageURL = (url?: string) =>
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/utils/string-utils.ts:8]: Can we add a test for this util? 🙏

Comment (yrliou) [reply components/brave_wallet_ui/utils/string-utils.ts:8]: https://github.com/brave/brave-core/pull/11518/commits/cbb07ff5299e7a022398272cc5ff9c3fd277bb41

**Review by Douglashdaniel — APPROVED** [review]
Front-end looks good!

File: components/brave_wallet_ui/utils/string-utils.ts (hunk 3 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/utils/string-utils.ts:8]: Thanks! Looks good!

**Comment by yrliou** [comment]
linux failed at s3-upload.

**Pull request merged**